File: PROYECTO ERP/frontend/src/api/ocr.ts
import { apiClient } from "./client";
import type { Paginated } from "@/lib/types";

export type EstadoOcr =
  | "pendiente"
  | "procesando"
  | "extraido"
  | "confirmado"
  | "descartado"
  | "error";

export type TipoComprobanteOcr =
  | "factura"
  | "remito"
  | "presupuesto"
  | "desconocido";

export interface OcrItemExtraido {
  descripcion: string;
  cantidad: number | string;
  unidad: string;
  precio_unitario: string;
  subtotal?: string;
  articulo_id_match?: number | null;
}

export interface ProveedorMatchOcr {
  id: number;
  codigo: string;
  razon_social: string;
  cuit?: string | null;
}

export interface FacturaResumenOcr {
  id: number;
  numero: number;
  punto_venta: number;
  tipo: string;
  total: string;
}

export interface ComprobanteOcr {
  id: number;
  estado: EstadoOcr;
  tipo_detectado: TipoComprobanteOcr;
  letra: string | null;
  confianza: string | null;

  archivo_path: string;
  archivo_size_bytes: number;
  archivo_mime: string;

  proveedor_nombre_raw: string | null;
  proveedor_cuit_raw: string | null;
  proveedor_id_match: number | null;
  proveedor_match: ProveedorMatchOcr | null;

  numero_comprobante: string | null;
  fecha_comprobante: string | null;

  subtotal: string | null;
  iva_total: string | null;
  total: string | null;

  items_extraidos: OcrItemExtraido[];
  error_message: string | null;

  uploaded_by_user_id: number;
  sucursal_id: number | null;
  factura_creada_id: number | null;
  factura_creada: FacturaResumenOcr | null;

  duracion_extraccion_ms: number | null;
  modelo_ia_usado: string | null;

  created_at: string;
  updated_at: string;
}

export interface OcrItemOverridePayload {
  descripcion: string;
  cantidad: string;
  unidad: string;
  precio_unitario: string;
  iva_porc?: string;
  descuento_porc?: string;
  articulo_id?: number | null;
  crear_articulo_si_falta?: boolean;
}

export interface OcrConfirmarPayload {
  sucursal_id: number;
  proveedor_id?: number | null;
  numero_override?: string | null;
  fecha_override?: string | null;
  observacion?: string | null;
  items: OcrItemOverridePayload[];
}

export async function uploadComprobante(
  file: File,
  sucursalId?: number,
): Promise<ComprobanteOcr> {
  const fd = new FormData();
  fd.append("file", file);
  if (sucursalId) fd.append("sucursal_id", String(sucursalId));
  const { data } = await apiClient.post<ComprobanteOcr>(
    "/ocr/comprobante",
    fd,
    {
      headers: { "Content-Type": "multipart/form-data" },
      timeout: 60_000,
    },
  );
  return data;
}

export async function listComprobantes(params: {
  estado?: EstadoOcr;
  page?: number;
  per_page?: number;
} = {}): Promise<Paginated<ComprobanteOcr>> {
  const { data } = await apiClient.get<Paginated<ComprobanteOcr>>(
    "/ocr/comprobantes",
    { params },
  );
  return data;
}

export async function getComprobante(id: number): Promise<ComprobanteOcr> {
  const { data } = await apiClient.get<ComprobanteOcr>(
    `/ocr/comprobantes/${id}`,
  );
  return data;
}

export function getImagenUrl(id: number): string {
  return `/api/v1/ocr/comprobantes/${id}/imagen`;
}

export async function confirmarComprobante(
  id: number,
  payload: OcrConfirmarPayload,
): Promise<ComprobanteOcr> {
  const { data } = await apiClient.post<ComprobanteOcr>(
    `/ocr/comprobantes/${id}/confirmar`,
    payload,
  );
  return data;
}

export async function descartarComprobante(
  id: number,
): Promise<ComprobanteOcr> {
  const { data } = await apiClient.post<ComprobanteOcr>(
    `/ocr/comprobantes/${id}/descartar`,
    {},
  );
  return data;
}

File: PROYECTO ERP/backend/app/schemas/ocr.py
"""Schemas Pydantic para OCR de comprobantes."""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.comprobante_ocr import EstadoOcrEnum, TipoComprobanteOcrEnum


class OcrItemExtraido(BaseModel):
    """Item extraído por la IA. Todos los campos son sugerencias editables."""

    model_config = ConfigDict(from_attributes=True)

    descripcion: str
    cantidad: Decimal = Field(default=Decimal("1"))
    unidad: str = Field(default="unidad")
    precio_unitario: Decimal = Field(default=Decimal("0"))
    subtotal: Decimal = Field(default=Decimal("0"))
    articulo_id_match: int | None = None


class ProveedorMatchOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    codigo: str
    razon_social: str
    cuit: str | None = None


class FacturaResumenOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    numero: int
    punto_venta: int
    tipo: str
    total: Decimal


class ComprobanteOcrOut(BaseModel):
    """Salida estándar del OCR."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    estado: EstadoOcrEnum
    tipo_detectado: TipoComprobanteOcrEnum
    letra: str | None = None
    confianza: Decimal | None = None

    archivo_path: str
    archivo_size_bytes: int
    archivo_mime: str

    proveedor_nombre_raw: str | None = None
    proveedor_cuit_raw: str | None = None
    proveedor_id_match: int | None = None
    proveedor_match: ProveedorMatchOut | None = None

    numero_comprobante: str | None = None
    fecha_comprobante: date | None = None

    subtotal: Decimal | None = None
    iva_total: Decimal | None = None
    total: Decimal | None = None

    items_extraidos: list[dict] = []
    error_message: str | None = None

    uploaded_by_user_id: int
    sucursal_id: int | None = None
    factura_creada_id: int | None = None
    factura_creada: FacturaResumenOut | None = None

    duracion_extraccion_ms: int | None = None
    modelo_ia_usado: str | None = None

    created_at: datetime
    updated_at: datetime


class OcrItemOverride(BaseModel):
    """Item editado por el usuario antes de confirmar."""

    descripcion: str
    cantidad: Decimal = Field(gt=0)
    unidad: str = Field(default="unidad")
    precio_unitario: Decimal = Field(ge=0)
    iva_porc: Decimal = Field(default=Decimal("21"), ge=0, le=100)
    descuento_porc: Decimal = Field(default=Decimal("0"), ge=0, le=100)
    articulo_id: int | None = None  # si null y crear_articulo_si_falta → crea uno
    crear_articulo_si_falta: bool = True


class OcrConfirmarPayload(BaseModel):
    sucursal_id: int
    proveedor_id: int | None = None
    numero_override: str | None = None
    fecha_override: date | None = None
    observacion: str | None = None
    items: list[OcrItemOverride] = Field(min_length=1)
